Fix for PlatformTarget not used in MSBuild runner

os-alan opened the pull request:
Implemented PlatformTarget specification in MSBuild arguments string. MSBuildRunner.cs simply doesn't have it.

Closes #332

UPD: set it to WIP state, to implement Patrik's suggestions (see comments)

patriksvensson commented on src/Cake.Common/Tools/MSBuild/MSBuildRunner.cs:
```diff
@@ -91,6 +98,17 @@ private ProcessArgumentBuilder GetArguments(FilePath solution, MSBuildSettings s
+            {
+                case PlatformTarget.MSIL:
+                    return "AnyCPU";                                    
+                default:
```
I think it would be good to handle the different cases explicitly and throw a `InvalidOperationException` here.

os-alan replied:
You think `InvalidOperartionException` would suitable here? Maybe `ArgumentOutOfRangeException` ?

patriksvensson replied:
I don't think that `ArgumentOutOfRangeException` fits well for this, since it's not an array type. Otherwise you could use `InvalidEnumArgumentException`.

patriksvensson commented on src/Cake.Common/Tools/MSBuild/MSBuildRunner.cs:
```diff
@@ -64,6 +64,13 @@ private ProcessArgumentBuilder GetArguments(FilePath solution, MSBuildSettings s
             }
 
+            // Buid for a specific platform?
+            if (settings.PlatformTarget != default(PlatformTarget))
```
Would be better if we made `PlatformTarget` nullable and only set the platform target if it's not null. This would be more in line with how we do it at other places.

os-alan replied:
I had same though originally, but explicitly specifying "AnyCPU" platform causes many other tests to fail (because before this change, AnyCPU was expected as default value) and I was too lazy to fix them. But, if you find this to be good idea as well, then it worth implementing :)

patriksvensson replied:
Ok, I understand what you mean.

The platform target is used to chose what version of MSBuild to be used, and this conflicts.
Since we used `MSIL` as the default platform target, tests are failing.

We could make the PlatformTarget nullable and replace the `GetAlternativeToolPaths` [MSBuildRunner.cs](https://github.com/cake-build/cake/blob/develop/src/Cake.Common/Tools/MSBuild/MSBuildRunner.cs#L146).

```csharp
protected override IEnumerable<FilePath> GetAlternativeToolPaths(MSBuildSettings settings)
{
    if (settings == null)
    {
        throw new ArgumentNullException("settings");
    }

    var platform = settings.PlatformTarget ?? PlatformTarget.MSIL;
    var path = MSBuildResolver.GetMSBuildPath(_fileSystem, _environment, settings.ToolVersion, platform);
    if (path != null)
    {
        return new[] { path };
    }

    return Enumerable.Empty<FilePath>();
}
```

We would also need to remove the assignment of the default PlatformTarget in the [MSBuildSettings constructor](https://github.com/cake-build/cake/blob/develop/src/Cake.Common/Tools/MSBuild/MSBuildSettings.cs#L80).

Finally, we should change (and rename) the [test for the default values in MSBuildSettings](https://github.com/cake-build/cake/blob/develop/src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildSettingsTests.cs#L22) to reflect these changes:

```csharp
[Fact]
public void Should_Set_Default_Platform_Target_To_Null()
{
    // Given, When
    var settings = new MSBuildSettings();

    // Then
    Assert.Null(settings.PlatformTarget);
}
```

What do you think?

os-alan replied:
That's exactly what I did :+1:

patriksvensson commented:
@os-alan Thanks for this! Not sure how `PlatformTarget` never was properly implemented but good that you found it! :+1: 

It looks good to me, but I left some minor comments that I think should be addressed before merging.

patriksvensson commented on src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs:
```diff
@@ -335,6 +335,27 @@ public void Should_Append_Configuration_As_Property_To_Process_Arguments()
             }
 
+            [Theory]
+            [InlineData(null, "/m /v:normal /target:Build \"/Working/src/Solution.sln\"")]
```
Create a new tests for this to avoid if statements in the test.

patriksvensson commented on src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildSettingsTests.cs:
```diff
@@ -18,16 +18,17 @@ public void Should_Set_Default_Tools_Version_To_Default()
 
                 // Then
-                Assert.Equal(PlatformTarget.MSIL, settings.PlatformTarget);
+                Assert.Equal(default(PlatformTarget?), settings.PlatformTarget);
```
I think `Assert.Null(settings.PlatformTarget)` would be more appropriate here.

patriksvensson commented on src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildSettingsTests.cs:
```diff
@@ -84,6 +85,20 @@ public void Should_Be_Empty_By_Default()
+                var settings = new MSBuildSettings();
+
+                // Then
+                Assert.Equal(default(PlatformTarget?), settings.PlatformTarget);
```
I think `Assert.Null(settings.PlatformTarget)` would be more appropriate here.

patriksvensson commented on src/Cake.Common/Tools/MSBuild/MSBuildRunner.cs:
```diff
@@ -150,7 +172,8 @@ protected override IEnumerable<FilePath> GetAlternativeToolPaths(MSBuildSettings
 
-            var path = MSBuildResolver.GetMSBuildPath(_fileSystem, _environment, settings.ToolVersion, settings.PlatformTarget);
+            var path = MSBuildResolver.GetMSBuildPath(_fileSystem, _environment, settings.ToolVersion,
+                                                       settings.PlatformTarget.HasValue ? settings.PlatformTarget.Value : PlatformTarget.MSIL);
```
Move this to it's own line for brevity like: `var platform = settings.PlatformTarget ?? PlatformTarget.MSIL`

patriksvensson commented:
Added some comments again :smile:

os-alan commented:
Fixed again :-)

patriksvensson commented on src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs:
```diff
@@ -335,6 +335,37 @@ public void Should_Append_Configuration_As_Property_To_Process_Arguments()
+                // Given
+                var fixture = new MSBuildRunnerFixture(false, true);
+                
+                if (platform.HasValue)
```
I think this if-statement should be omitted.

patriksvensson commented on src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs:
```diff
@@ -335,6 +335,37 @@ public void Should_Append_Configuration_As_Property_To_Process_Arguments()
+            {
+                // Given
+                var fixture = new MSBuildRunnerFixture(false, true);
+                // When
```
Line break

patriksvensson commented on src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildSettingsTests.cs:
```diff
@@ -84,6 +85,20 @@ public void Should_Be_Empty_By_Default()
+            }
+        }
+
+
```
Remove extra line break

patriksvensson commented on src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildSettingsTests.cs:
```diff
@@ -18,16 +18,17 @@ public void Should_Set_Default_Tools_Version_To_Default()
+                Assert.Null(settings.PlatformTarget);
             }
 
+
```
Remove extra line break

patriksvensson commented:
@os-alan Looks good to me. Some final minor comments and we're good to merge.

Would also be nice if you could rebase your work to one commit with a commit message that's more in line with others such as "Fix for PlatformTarget not used in MSBuild runner" or similar.

Great work! :+1:

os-alan commented:
Hi Patrik! 

>Would also be nice if you could rebase your work to one commit with a commit message that's more in line with others such as "Fix for PlatformTarget not used in MSBuild runner" or similar.

Fixed formatting and rebased. Hope, it will pass now :sweat_smile:

patriksvensson commented:
Hello @os-alan!

Yes, everything looks good now and I will merge this later tonight! :+1: 

/ Patrik

patriksvensson commented:
Merged :+1:
